File: client/fedcampus/src/main/java/org/eu/fedcampus/train/Train.kt
package org.eu.fedcampus.train

import android.content.Context
import android.util.Log
import io.grpc.ManagedChannel
import kotlinx.coroutines.*
import org.eu.fedcampus.train.db.TFLiteModel
import retrofit2.http.*
import java.io.File
import java.nio.MappedByteBuffer
import kotlin.properties.Delegates

class Train<X : Any, Y : Any> constructor(
    val context: Context,
    backendUrl: String,
    val sampleSpec: SampleSpec<X, Y>,
) {
    var sessionId: Int? = null
    var telemetry = false
        private set
    var deviceId by Delegates.notNull<Long>()
        private set
    val client = HttpClient(backendUrl)
    var state: TrainState<X, Y> = TrainState.Initialized()

    fun enableTelemetry(id: Long) {
        deviceId = id
        telemetry = true
    }

    @Suppress("unused")
    fun disableTelemetry() {
        deviceId = 0
        telemetry = false
    }

    /**
     * Download advertised model information.
     */
    @Throws
    suspend fun advertisedModel(dataType: String): TFLiteModel = when (state) {
        is TrainState.Initialized, is TrainState.WithModel -> doAdvertisedModel(dataType)
        else -> throw IllegalStateException("`advertisedModel` called with $state")
    }

    private suspend fun doAdvertisedModel(dataType: String): TFLiteModel {
        val model = client.advertisedModel(PostAdvertisedData(dataType))
        Log.d(TAG, "Model: $model")
        state = TrainState.WithModel(model)
        return model
    }

    @Suppress("UNUSED_PARAMETER")
    @Throws
    fun modelDownloaded(model: TFLiteModel): Boolean {
        // TODO: Save model to DB.
        return false
    }

    /**
     * Download TFLite files to `"models/$path"` if they have not been saved to DB.
     */
    @Throws
    suspend fun downloadModelFile(modelDir: File): File = state.let {
        when (it) {
            is TrainState.WithModel -> doDownloadModelFile(modelDir, it.model)
            else -> throw IllegalStateException("`downloadModelFile` called with $state")
        }
    }

    private suspend fun doDownloadModelFile(modelDir: File, model: TFLiteModel): File {
        val fileUrl = model.file_path
        val fileName = fileUrl.split("/").last()
        if (modelDownloaded(model)) {
            // The model is already in the DB
            Log.i(downloadModelFileTag, "skipping already downloaded model ${model.name}")
            return File(modelDir, fileName)
        }
        val fileDir = client.downloadFile(fileUrl, modelDir, fileName)
        Log.i(downloadModelFileTag, "$fileUrl -> ${fileDir.absolutePath}")
        return fileDir
    }

    @Throws
    suspend fun getServerInfo(start_fresh: Boolean = false): ServerData = state.let {
        when (it) {
            is TrainState.WithModel -> doGetServerInfo(it.model, start_fresh)
            else -> throw IllegalStateException("`getServerInfo` called with $state")
        }
    }

    private suspend fun doGetServerInfo(model: TFLiteModel, start_fresh: Boolean): ServerData {
        val serverData = client.postServer(model, start_fresh)
        sessionId = serverData.session_id
        Log.i(TAG, "Server data: $serverData")
        return serverData
    }

    /**
     * Ask backend for advertised model, load model into [state], and download its corresponding file.
     * @return Model file.
     */
    @Throws
    suspend fun prepareModel(dataType: String): File = when (state) {
        is TrainState.Initialized, is TrainState.WithModel -> doPrepareModel(dataType)
        else -> throw IllegalStateException("`prepareModel` called with $state")
    }

    private suspend fun doPrepareModel(dataType: String) =
        withContext(Dispatchers.IO) {
            val model = advertisedModel(dataType)
            val modelDir = model.getModelDir(context)
            downloadModelFile(modelDir)
        }

    /**
     * Initialize Flower Client with TFLite model [buffer] and establish channel connection to Flower server.
     */
    @Throws
    suspend fun prepare(buffer: MappedByteBuffer, address: String, useTLS: Boolean) = state.let {
        when (it) {
            is TrainState.WithModel -> doPrepare(buffer, address, useTLS, it.model)
            else -> throw IllegalStateException("`prepare` called with $state")
        }
    }

    private suspend fun doPrepare(
        buffer: MappedByteBuffer,
        address: String,
        useTLS: Boolean,
        model: TFLiteModel
    ): FlowerClient<X, Y> {
        val flowerClient = FlowerClient(buffer, model, sampleSpec)
        val channel = withContext(Dispatchers.IO) {
            createChannel(address, useTLS)
        }
        state = TrainState.Prepared(model, flowerClient, channel)
        return flowerClient
    }

    /**
     * Only call this after loading training data into the Flower Client.
     */
    @Throws
    fun start(callback: (String) -> Unit) = state.let {
        when (it) {
            is TrainState.Prepared -> doStart(callback, it.model, it.flowerClient, it.channel)
            else -> throw IllegalStateException("`start` called with $state")
        }
    }

    @Throws
    private fun doStart(
        callback: (String) -> Unit,
        model: TFLiteModel,
        flowerClient: FlowerClient<X, Y>,
        channel: ManagedChannel
    ) =
        if (flowerClient.trainingSamples.isEmpty() || flowerClient.testSamples.isEmpty()) {
            throw Error("No data loaded for training")
        } else {
            FlowerServiceRunnable(
                channel,
                this,
                model,
                flowerClient,
                callback
            ).let {
                state = TrainState.Training(model, flowerClient, it)
                it
            }
        }

    /**
     * Ensure that telemetry is enabled and [sessionId] is non-null.
     */
    @Throws(AssertionError::class)
    fun checkTelemetryEnabled() {
        assert(telemetry)
        assert(sessionId !== null)
    }

    @Throws
    suspend fun fitInsTelemetry(start: Long, end: Long) {
        checkTelemetryEnabled()
        val body = FitInsTelemetryData(deviceId, sessionId!!, start, end)
        client.fitInsTelemetry(body)
        Log.i(TAG, "Telemetry: Sent fit instruction telemetry")
    }

    @Throws
    suspend fun evaluateInsTelemetry(
        start: Long,
        end: Long,
        loss: Float,
        accuracy: Float,
        test_size: Int
    ) {
        checkTelemetryEnabled()
        val body =
            EvaluateInsTelemetryData(deviceId, sessionId!!, start, end, loss, accuracy, test_size)
        client.evaluateInsTelemetry(body)
        Log.i(TAG, "Telemetry: Sent evaluate instruction telemetry")
    }

    companion object {
        const val TAG = "Train"
        const val downloadModelFileTag = "Download TFLite model"
    }
}

File: flutter/lib/train.dart
import 'dart:io';

import 'package:fed_kit/backend_client.dart';
import 'package:fed_kit/flower_service.dart';
import 'package:fed_kit/log.dart';
import 'package:fed_kit/ml_client.dart';
import 'package:fed_kit/ml_model.dart';
import 'package:fed_kit/train_state.dart';
import 'package:grpc/grpc.dart';

class Train {
  bool _telemetry = false;
  bool get telemetry => _telemetry;
  int _deviceId = 0;
  int get deviceId => _deviceId;
  late BackendClient _client;
  TrainState _state = Initialized();
  int? _sessionId;

  Train(String backendUrl) {
    _client = BackendClient(backendUrl);
  }

  void enableTelemetry(int deviceId) {
    _telemetry = true;
    _deviceId = deviceId;
  }

  void disableTelemetry() {
    _deviceId = 0;
    _telemetry = false;
  }

  Future<(MLModel, String)> prepareModel(String dataType) => switch (_state) {
        Initialized() || WithModel() => _prepareModel(dataType),
        _ => throw Exception('`prepareModel` called with $_state'),
      };

  Future<(MLModel, String)> _prepareModel(String dataType) async {
    final model = await _whichModel(dataType);
    final (modelUrl, modelDir) = await model.urlAndDir;
    await downloadModelFile(modelUrl, modelDir);
    return (model, modelDir);
  }

  Future<MLModel> whichModel(String dataType) async => switch (_state) {
        Initialized() || WithModel() => _whichModel(dataType),
        _ => throw Exception('`advertisedModel` called with $_state'),
      };

  Future<MLModel> _whichModel(String dataType) async {
    final model = await _client.whichModel(PostAdvertisedData(
        data_type: dataType, tflite: !Platform.isIOS, coreml: Platform.isIOS));
    logger.d('Advertised model: $model.');
    _state = WithModel(model);
    return model;
  }

  Future<void> downloadModelFile(String modelUrl, String modelDir) =>
      switch (_state) {
        WithModel state => _downloadModelFile(modelUrl, modelDir, state.model),
        _ => throw Exception('`downloadModelFile` called with $_state'),
      };

  Future<void> _downloadModelFile(
      String modelUrl, String modelDir, MLModel model) async {
    if (await Directory(modelDir).exists()) {
      logger.d('Skipping already downloaded model ${model.name}');
      return;
    }
    await _client.downloadFile(modelUrl, modelDir);
    logger.d('Downloaded ${model.name}: $modelUrl -> $modelDir.');
  }

  Future<ServerData> getServerInfo({bool startFresh = false}) =>
      switch (_state) {
        WithModel state => _getServerInfo(state.model, startFresh),
        _ => throw Exception('`getServerInfo` called with $_state'),
      };

  Future<ServerData> _getServerInfo(MLModel model, bool startFresh) async {
    final serverData = await _client.postServer(model, startFresh);
    _sessionId = serverData.session_id;
    logger.d('Server info: $serverData.');
    return serverData;
  }

  /// After ML client is ready.
  Future<void> prepare(MLClient mlClient, String address, int port,
          {ChannelOptions channelOptions = defaultChannelOptions}) =>
      switch (_state) {
        WithModel state =>
          _prepare(mlClient, state, address, port, channelOptions),
        _ => throw Exception('`mlClientReady` called with $_state'),
      };

  Future<void> _prepare(MLClient mlClient, WithModel state, String address,
      int port, ChannelOptions options) async {
    if (!await mlClient.ready()) {
      throw Exception('`mlClient` not ready');
    }
    final channel = ClientChannel(
      address,
      port: port,
      options: options,
    );

    _state = Prepared(state.model, mlClient, channel);
  }

  Stream<String> start() => switch (_state) {
        Prepared state => _start(state),
        _ => throw Exception('`start` called with $_state'),
      };

  Stream<String> _start(Prepared state) {
    final model = state.model;
    final flowerService =
        FlowerService(state.channel, this, state.mlClient, model);
    final stream = flowerService.run();
    logger.d('Training for ${model.name} started.');
    _state = Training(model, flowerService);
    return stream;
  }

  Future<void> fitInsTelemetry(DateTime start, DateTime end) async {
    checkTelemetryEnabled();
    final body = FitInsTelemetryData(
        device_id: deviceId,
        session_id: _sessionId!,
        start: start.millisecondsSinceEpoch,
        end: end.millisecondsSinceEpoch);
    await _client.fitInsTelemetry(body);
    logger.d('Train telemetry: sent FitIns.');
  }

  Future<void> evaluateInsTelemetry(DateTime start, DateTime end, double loss,
      double accuracy, int testSize) async {
    checkTelemetryEnabled();
    final body = EvaluateInsTelemetryData(
        device_id: deviceId,
        session_id: _sessionId!,
        start: start.millisecondsSinceEpoch,
        end: end.millisecondsSinceEpoch,
        loss: loss,
        accuracy: accuracy,
        test_size: testSize);
    await _client.evaluateInsTelemetry(body);
    logger.d('Train telemetry: sent EvaluateIns.');
  }

  void checkTelemetryEnabled() {
    if (!telemetry || _sessionId == null) {
      throw Exception('Telemetry disabled');
    }
  }
}

const defaultChannelOptions =
    ChannelOptions(credentials: ChannelCredentials.insecure());
